pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java

```diff
@@ -0,0 +1,507 @@
+        // https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#taint-based-evictions
+        podSpec.setTolerations(getTolerations());
+
+        podSpec.containers(Collections.singletonList(
+                getContainer(instanceCommand, resource)));
+
+        return podSpec;
+    }
+
+    private List<V1Toleration> getTolerations() {
+        final List<V1Toleration> tolerations = new ArrayList<>();
+        TOLERATIONS.forEach(t -> {
+            final V1Toleration toleration =
+                    new V1Toleration()
+                            .key(t)
+                            .operator("Exists")
+                            .effect("NoExecute")
+                            .tolerationSeconds(10L);
+            tolerations.add(toleration);
+        });
+
+        return tolerations;
+    }
+
+    private V1Container getContainer(List<String> instanceCommand, Function.Resources resource) {
+        final V1Container container = new V1Container().name("pulsarfunction");
+
+        // set up the container images
+        container.setImage(pulsarDockerImageName);
+
+        // set up the container command
+        container.setCommand(instanceCommand);
+
+        // setup the environment variables for the container
+        final V1EnvVar envVarPodName = new V1EnvVar();
+        envVarPodName.name("POD_NAME")
+                .valueFrom(new V1EnvVarSource()
+                        .fieldRef(new V1ObjectFieldSelector()
+                                .fieldPath("metadata.name")));
+        container.setEnv(Arrays.asList(envVarPodName));
+
+
+        // set container resources
+        final V1ResourceRequirements resourceRequirements = new V1ResourceRequirements();
+        final Map<String, Quantity> requests = new HashMap<>();
+        requests.put("memory", Quantity.fromString(Long.toString(resource != null && resource.getRam() != 0 ? resource.getRam() : 1073741824)));
+        requests.put("cpu", Quantity.fromString(Double.toString(resource != null && resource.getCpu() != 0 ? resource.getCpu() : 1)));
+        resourceRequirements.setRequests(requests);
+        container.setResources(resourceRequirements);
+
+        // set container ports
+        container.setPorts(getContainerPorts());
+
+        return container;
+    }
+
+    private List<V1ContainerPort> getContainerPorts() {
+        List<V1ContainerPort> ports = new ArrayList<>();
+        final V1ContainerPort port = new V1ContainerPort();
+        port.setName("grpc");
+        port.setContainerPort(GRPC_PORT);
+        ports.add(port);
+        return ports;
+    }
+
+    private static String createJobName(Function.FunctionDetails functionDetails) {
+        return createJobName(functionDetails.getTenant(),
+                functionDetails.getNamespace(),
+                functionDetails.getName());
+    }
+
+    private static String createJobName(String tenant, String namespace, String functionName) {
```

Comment:
I've changed this to include tenant/namespace name as well

Reply:
Should we not do some short of mapping of tenant/namespace to some kubernetes namespace?

Reply:
Not sure about that. Job namespace in Kubernetes is more involved in the sense that someone needs to create them and give perms for pulsar worker to be able to submit to them.
In any case, we might want to decide these kind of functionality based on usage rather than add it in the first cut.